packages/apps/contracts/PreimageApp.sol

```diff
@@ -0,0 +1,70 @@
+pragma solidity 0.5.8;
+pragma experimental "ABIEncoderV2";
+
+import "@counterfactual/contracts/contracts/interfaces/CounterfactualApp.sol";
+
+
+contract PreimageApp {
+
+  struct AppState {
+    bytes32 image;
+    bool preimageRevealed;
+  }
+
+  struct Action {
+    bytes preimage;
+  }
+
+  function isStateTerminal(bytes memory encodedState)
+    public
+    pure
+    returns (bool)
+  {
+    AppState memory state = abi.decode(encodedState, (AppState));
+    return state.preimageRevealed;
+  }
+
+  function getTurnTaker(
+    bytes memory /* encodedState */, address[] memory signingKeys
+  )
+    public
+    pure
+    returns (address)
+  {
+    return signingKeys[0];
+  }
+
+  function applyAction(bytes memory encodedState, bytes memory encodedAction)
```

Comment:
Why do you use `memory` sometimes and `calldata` other times? When we call these methods from the `Node` it will be `calldata` but when we call from `AppRegistry` it will be `memory`... I think? Or is it always `calldata`? In any case, this is the kind of standardization we should be getting from inheritance, no?

Reply:
When public functions are called, the params passed in calldata are copied into memory IIRC. We should use `external/calldata` where possible

Reply:
OK, can you change all instances to use `external` and `calldata` then?

Reply:
https://github.com/counterfactual/monorepo/issues/1449